src/pymedphys/dicom/_level3/dicom_collection.py

```diff
@@ -0,0 +1,111 @@
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version (the "AGPL-3.0+").
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License and the additional terms for more
+# details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# ADDITIONAL TERMS are also included as allowed by Section 7 of the GNU
+# Affero General Public License. These additional terms are Sections 1, 5,
+# 6, 7, 8, and 9 from the Apache License, Version 2.0 (the "Apache-2.0")
+# where all references to the definition "License" are instead defined to
+# mean the AGPL-3.0+.
+
+# You should have received a copy of the Apache-2.0 along with this
+# program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+
+
+from copy import deepcopy
+
+import pydicom
+
+from .._level1.create import dicom_dataset_from_dict
+from .._level2.anonymise import anonymise_dicom_dataset
+from .._level2.dose import (
+    extract_dicom_patient_xyz, convert_xyz_to_dicom_coords)
+
+from ...libutils import get_imports
+
+IMPORTS = get_imports(globals())
+
+# pylint: disable=W0201
+
+
+class DicomBase:
+    def __init__(self, dataset):
+        self.dataset = dataset
+
+    @property
+    def dataset(self) -> pydicom.Dataset:
+        return deepcopy(self._dataset)
+
+    @dataset.setter
+    def dataset(self, dataset: pydicom.Dataset):
+        self._dataset = deepcopy(dataset)
+        self.after_assigning_dataset()
+
+    def after_assigning_dataset(self):
+        pass
+
+    @classmethod
+    def from_file(cls, filepath):
+        dataset = pydicom.read_file(filepath, force=True)
+
+        return cls(dataset)
+
+    @classmethod
+    def from_dict(cls, dictionary):
+        dataset = dicom_dataset_from_dict(dictionary)
+
+        return cls(dataset)
+
+    def __repr__(self):
+        return self._dataset.__repr__()
+
+    def save_as(self, filepath):
+        self._dataset.save_as(filepath)
+
+    def anonymise(self):
+        anonymised = anonymise_dicom_dataset(self.dataset)
+
+        return self.__class__(anonymised)
+
+
+class DicomDose(DicomBase):
+    def after_assigning_dataset(self):
+        self.values = self._dataset.pixel_array * self._dataset.DoseGridScaling
+        self.units = self._dataset.DoseUnits
+
+        self.x, self.y, self.z = extract_dicom_patient_xyz(self._dataset)
+        self.coords = convert_xyz_to_dicom_coords((self.x, self.y, self.z))
+        self.mask = None
+
+
+class DicomCT(DicomBase):
+    pass
+
+
+class DicomStructure(DicomBase):
+    pass
+
+
+class DicomPlan(DicomBase):
+    pass
+
+
+class DicomCollection:
```

Comment:
In the interest of maintaining nomenclature parallel to the DICOM Standard, would this correspond to a Study (implying that the class be named `DicomStudy`? See http://www.ssrpm.ch/old/dicom/neumann1.pdf

Maybe not a perfect correspondence, since I think DICOM CT *strictly* lies outside of an RT Study. But the CT is implicit in the Structure Set, I believe. So I think it's still a pretty good match.

Reply:
Yes, I agree, and maybe we need an extra layer or two. We should possibly have a DicomImage, and maybe instead of DicomCT we should have DicomSeries which would be a series of CT images. It would be worth thinking of whether or not it might be worthwhile adjusting how these pixel values are stored. Let's think on that when we're trying to use it, after it is all made.

We should also have a DicomView class which doesn't actually copy the data across but just acta as a pointer to another DicomObject.

A DicomStudy would be an inter-related  single Series, Plan, Structure, Dose. We need to put some work here to be memory efficient. Allow DicomStudy to contain the objects themselves or simply views to the objects.

A DicomCollection would still exist, but it would allow an arbitrary number of Dicom Studies within it. Any Dicom object with the same UID as another would be verified for equality, an error raised if not the same Dicom dataset, and then a view would be created instead of another full object. That way looking up each Dicom study in the collection would still give all the info, and still look like that CT series is in every DicomStudy in the collection. As well as editing the CT slices on one study in the collection should edit the slices on every other study in that collection also.

Multiple different doses would look like multiple different studies, but the CT series, structures, plan would all be a shared view between them.

... to make this simpler maybe we shouldn't do that defensive copy...

Reply:
If we don't make a defensive copy then we won't need to distinguish between view and otherwise.